modules/storage/storage.go

```diff
@@ -0,0 +1,287 @@
+			if err != nil {
+				return err
+			}
+
+			newPools = append(newPools, pool)
+			createdPools++
+		}
+	}
+
+	// make sure new pools are mounted
+	log.Info().Msgf("Making sure new volumes are mounted")
+	for _, pool := range newPools {
+		if _, mounted := pool.Mounted(); !mounted {
+			log.Debug().Msgf("Mounting volume %s", pool.Name())
+			if _, err = pool.Mount(); err != nil {
+				return err
+			}
+		}
+	}
+
+	s.volumes = append(s.volumes, newPools...)
+
+	return s.ensureCache()
+}
+
+// CreateFilesystem with the given size in a storage pool.
+func (s *storageModule) CreateFilesystem(size uint64) (string, error) {
+	log.Info().Msgf("Creating new volume with size %d", size)
+
+	fs, err := s.createFs(size, uuid.New().String(), false)
+	if err != nil {
+		return "", err
+	}
+	return fs.Path(), nil
+}
+
+// ReleaseFilesystem at the given path, this will unmount and then delete
+// the filesystem. After this call, the caller must not perform any more actions
+// on this filesystem
+func (s *storageModule) ReleaseFilesystem(path string) error {
+	log.Info().Msgf("Deleting volume at %v", path)
+
+	for _, volume := range s.volumes {
+		filesystems, err := volume.Volumes()
+		if err != nil {
+			return err
+		}
+		for _, fs := range filesystems {
+			if fs.Path() == path {
+				log.Debug().Msgf("Removing filesystem %v in volume %v", fs.Name(), volume.Name())
+				return volume.RemoveVolume(fs.Name())
+			}
+		}
+	}
+
+	log.Warn().Msgf("Could not find filesystem %v", path)
+	return nil
+}
+
+// ensureCache creates a "cache" subvolume and mounts it in /var
+func (s *storageModule) ensureCache() error {
+	log.Info().Msgf("Setting up cache")
+
+	log.Debug().Msgf("Checking pools for existing cache")
+
+	var cacheFs filesystem.Volume
+
+	// check if we already have a cache
+	for _, volume := range s.volumes {
+		filesystems, err := volume.Volumes()
+		if err != nil {
+			return err
+		}
+		for _, fs := range filesystems {
+			if fs.Name() == cacheLabel {
+				log.Debug().Msgf("Found existing cache at %v", fs.Path())
+				cacheFs = fs
+				break
+			}
+		}
+		if cacheFs != nil {
+			break
+		}
+	}
+
+	if cacheFs == nil {
+		log.Debug().Msgf("No cache found, try to create new cache")
+
+		fs, err := s.createFs(cacheSize, cacheLabel, true)
+		if err != nil {
+			return err
+		}
+		cacheFs = fs
+	}
+
+	log.Debug().Msgf("Mounting cache partition in %s", cacheTarget)
+	return filesystem.BindMount(cacheFs, cacheTarget)
+}
+
+// createFs creates a filesystem with the given name and limits it to the given size
+// if the requested disk type does not have a storage pool available, an error is
+// returned
+func (s *storageModule) createFs(size uint64, name string, preferSSD bool) (filesystem.Volume, error) {
+	var fs filesystem.Volume
+	var err error
+
+	possiblePools := s.volumes
+	if preferSSD {
+		ssdPools := []filesystem.Pool{}
+		hddPools := []filesystem.Pool{}
+		for _, pool := range s.volumes {
```

Comment:
Can't we avoid to go back to the disks to get the PoolType ?
I would like we try to let the disks alone as much as possible.
So when the storage module boots, it does a scanning of all the disks. Then keeps the required info in memory, so we can completely turn off all the unused disk.

Reply:
it's a compromise between the following:
- Caching
- Performance
- Number of operations

Caching, is usually problematic, and can go easily out of sync. I already hate that we keep the available pools in memory, since it can go out of sync with the system if another cache (some how) was created or delete behind the storage module back.

Performance is a good concern, going to the disk is indeed much slower than checking some cached in memory value, but this brings us to the last point.

Number of operations, basically how often do we need to do this ? it's only on creating a new subvolume which i think is not that many operations per hours. 

I would actually recommend we drop all memory state all together, and make sure everything is always checked during runtime.

Reply:
Regarding turning off all unused disks, once the disk is mounted we can't turn it off. Only unformatted disks can be safely turned off. and this can be done only one time on storage module booting.

Then on creating new pool, we will have to check all "free" disks to get the type. then create the pool, and turn the remaining disks off again.

Reply:
The actual problem with the current code, is that we don't cache the disk type. And since we actually do some write to the disk to determine its type (seektime) then every time we create a new pool, we have to turn on the disk, execute seektime. This is quite redundant since the type of a disk will never change, Once you know it, you know it forever.

I agree with you that caching is a bitch and I would like to avoid it too.
Keeping the pools in memory doesn't really bring anything here since the type is not saved.

> since it can go out of sync with the system if another cache (some how) was created or delete behind the storage module back.

Regarding this, the real problem is that if we go out of sync, that means or we have a bug in our code, or the node is compromised. Since the storage module MUST be the only one dealing with disks...

Reply:
seektime does not write to the disk. it only reads random blocks and calculate the time between the seeks to the random position. Just saying.

But anyway, since we generate the pools names, we can save some data in the pool label to extract such information without actually storing it anywhere.

so a pool label can be (instead of only a guid) be in a format like `type:id` instead. so pool type basically can be extracted from name.

@LeeSmet what do u think ?

Reply:
> seektime does not write to the disk. it only reads random blocks and calculate the time between the seeks to the random position. Just saying.

Correct, but my point stays, the disk need to be active for that.

> so a pool label can be (instead of only a guid) be in a format like type:id instead. so pool type basically can be extracted from name.

That's a nice idea actually. That would avoid caching and solve the type issue at the same time

Reply:
Imo, if we want the storage module to be the only thing managing disks, then caching should be fine . If the cached info is out of date with the disks that can mean 2 things: a bug in the storage module (forgot to rescan disks after modification), or something else performed an action on the disks, which is then also considered a bug as storage should be the only thing doing that. Labeling without caching would mean we need to scan for filesystems every time, does this not have a performance impact?